packages/kit/src/runtime/app/server/remote.js

```diff
@@ -0,0 +1,827 @@
+ */
+/**
+ * @template Input
+ * @template Output
+ * @param {any} validate_or_fn
+ * @param {any} [fn_or_options]
+ * @param {{ inputs?: RemotePrerenderInputsGenerator<Input>, dynamic?: boolean }} [maybe_options]
+ * @returns {RemoteQuery<Input, Output>}
+ */
+/*@__NO_SIDE_EFFECTS__*/
+export function prerender(validate_or_fn, fn_or_options, maybe_options) {
+	check_experimental('prerender');
+
+	const maybe_fn = typeof fn_or_options === 'function' ? fn_or_options : undefined;
+	/** @type {typeof maybe_options} */
+	const options = maybe_options ?? (maybe_fn ? undefined : fn_or_options);
+	/** @type {(arg?: Input) => Output} */
+	const fn = maybe_fn ?? validate_or_fn;
+	/** @type {(arg?: any) => MaybePromise<Input>} */
+	const validate = create_validator(validate_or_fn, maybe_fn);
+
+	/** @type {RemoteQuery<Input, Output> & { __: RemoteInfo }} */
+	const wrapper = (arg) => {
+		/** @type {Partial<ReturnType<RemoteQuery<Input, Output>>>} */
+		const promise = (async () => {
+			const event = getRequestEvent();
+			const info = get_remote_info(event);
+			const stringified_arg = stringify_remote_arg(arg, info.transport);
+			const id = wrapper.__.id;
+			const url = `${base}/${app_dir}/remote/${id}${stringified_arg ? `/${stringified_arg}` : ''}`;
+
+			if (!info.prerendering && !DEV && !event.isRemoteRequest) {
+				try {
+					return await get_response(id, arg, event, async () => {
+						const response = await fetch(event.url.origin + url);
+						if (!response.ok) {
+							throw new Error('Prerendered response not found');
+						}
+						const prerendered = await response.json();
+						info.results[create_remote_cache_key(id, stringified_arg)] = prerendered.result;
+						return parse_remote_response(prerendered.result, info.transport);
+					});
+				} catch {
+					// not available prerendered, fallback to normal function
+				}
+			}
+
+			if (info.prerendering?.remote_responses.has(url)) {
+				return /** @type {Promise<any>} */ (info.prerendering.remote_responses.get(url));
+			}
+
+			const maybe_promise = get_response(id, arg, event, () =>
+				run_remote_function(event, false, arg, validate, fn)
+			);
+
+			if (info.prerendering) {
+				info.prerendering.remote_responses.set(url, Promise.resolve(maybe_promise));
+				Promise.resolve(maybe_promise).catch(() => info.prerendering?.remote_responses.delete(url));
+			}
+
+			const result = await maybe_promise;
+
+			if (info.prerendering) {
+				const body = { type: 'result', result: stringify(result, info.transport) };
+				info.prerendering.dependencies.set(url, {
+					body: JSON.stringify(body),
+					response: json(body)
+				});
+			}
+
+			return result;
+		})();
+
+		promise.refresh = () => {
+			throw new Error(
+				`Cannot call  '${wrapper.__.name}.refresh()'. Remote prerender functions are immutable and cannot be refreshed.`
+			);
+		};
+
+		promise.override = () => {
+			throw new Error(`Cannot call '${wrapper.__.name}.override()' on the server`);
+		};
```

Comment:
this feels like a type issue. `query` and `prerender` return values shouldn't share a type